### ISSUE TITLE
[Improvement] send proper WebSocket close code and reason during shutdown of the client and server

### ISSUE THREAD

**Issue by Fetsivalen** [reactor/reactor-netty#1091]
During close web socket server\client currently send WebSocketCloseStatus with code -1.
It seems to be more logical to have default value `1001` (meaning server going away) according to the https://tools.ietf.org/html/rfc6455#section-7.4 . 

## Motivation
Using standard closed codes during server/client shutdown will allow "the other side of communication" to react properly on them.

## Desired solution
Change default close code and reason during the shutdown of the server/client to 1001. And make it configurable. For example,  it can be useful to use not RFC code (1012 for example).

## Considered alternatives
There could be a few alternatives.
1 keep default behavior as is (to keep backward compatibility) and allow to configure desired closed code.
2. Use another default closed code.

## Additional context
None

**Comment by violetagg** [comment]
@Fetsivalen Is it possible in your use case to use one of the `sendClose` methods?
https://projectreactor.io/docs/netty/release/api/reactor/netty/http/websocket/WebsocketOutbound.html

**Comment by Fetsivalen** [comment]
My case is the same as the original one in Spring Cloud Gateway.
From what I see the only way to use one of the `sendClose` methods is to add  (onTerminate?)  hooks to both sides of the communication.
Here is an example in Spring cloud Gateway (which is actually my case)
https://github.com/spring-cloud/spring-cloud-gateway/blob/master/spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/filter/WebsocketRoutingFilter.java#L186-L194
I'm going to try this in my code but it seems to me that it would be useful to have it out of the box in reactor-netty.
Do you think that this change should be in Spring Cloud Gateway, not here?

**Comment by violetagg** [comment]
@rstoyanchev @spencergibb We have API for sending a custom close frame but I cannot see how this API can be used in Spring Gateway use case. Can you please take a look?

**Comment by violetagg** [comment]
@Fetsivalen According to Spring Framework javadoc you can call close on the session
https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/reactive/socket/WebSocketSession.html#close--
I think this can be used. 
Wdyt?

**Comment by Fetsivalen** [comment]
Hi @violetagg 
Thanks for the suggestion
Actually I've reworked a bit Spring Cloud Gateway's implementation (will open PR once it will reach production)
So in my case, when Gateway bypassed closed code and reason from client to the server and visa versa this method definitely works great. (omitted metrics and my custom code in the example).
```
 Mono<Void> clientClosedStatusStream = clientSession.receiveClosedStatus()
                .flatMap(closedStatus -> serverSession.close(new CloseStatus(closedStatus.code(), closedStatus.reasonText())));

        Mono<Void> serverClosedStatusStream = serverSession.receiveClosedStatus()
                .flatMap(closedStatus -> clientSession.close(new CloseStatus(closedStatus.code(), closedStatus.reasonText())));

        Mono<Void> clientEventStream = serverSession.send(
                clientSession.receive()
                        .doOnNext(WebSocketMessage::retain)
        );

        Mono<Void> serverEventStream = clientSession.send(
                serverSession.receive()
                        .doOnNext(WebSocketMessage::retain)
        );

        return Mono.zip(clientClosedStatusStream, serverClosedStatusStream, clientEventStream, serverEventStream)
                .then();
```
But the use-case I've raised here is a bit different.
When I'm re-deploying the gateway service itself, zipped Mono is terminated and I can only execute hooks like [onTerminate](https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Mono.html#doOnTerminate-java.lang.Runnable-)
But it seems that its too late stage to pass the close frame to the client and the server.

I've played around receiveClosedStatus and different other hooks but miss the point where I can execute `close` operation on the application termination and properly pass going away status to the client and backend server from the Gateway.
Am I miss smth?

**Comment by violetagg** [comment]
@Fetsivalen 
> When I'm re-deploying the gateway service itself, zipped Mono is terminated and I can only execute hooks like onTerminate

How will you distinguish between a normal shutdown and accidental shutdown?

**Comment by Fetsivalen** [comment]
I'm not sure that I got the context you ask this question with.
But I'll try to explain a few flows I see.

When you're closing connection intentionally you should send close CloseFrame manually according to the use case. 
 > existing `close` method matches this use case

When service is restarting\shutting down based on some automation redeploy flow or because of some liveness prob in Kubernetes (or similar in other platforms), you should notify another side of communication that you're going away (meaning send CloseFrame on terminate event).
> right now in any case when the connection is not closed intentionally -1 code will be sent (and this case is exactly what I propose to change).

From what I understand client doesn't care why are you restarting, it just needs to know that to proceed with communication web-socket connection should be established again.

To make it even more configurable could be exposed method to register `onTerminate` hook which should determine actual use case and the exact code to send ClosedFrame with (for ex pass function as an input parameter). But from my use cases, I don't see a room to use it and i'd rather skip it as it may "not gonna needed".

Also, there are possible situations like `kill -9` when the process ends abnormally, but it seems to be out of scope here since hooks won't be executed in this case anyway. (from what I test most of the clients treats this case in the same way as "-1" code)
 
does it make sense?

A bit context that all my use cases are about the environment with multiple instances of each service, so "service going away" means for the client that it should just use "round-robin strategy again" to determine the instance to connect to.

**Comment by violetagg** [comment]
@Fetsivalen 
> When service is restarting\shutting down based on some automation redeploy flow or because of some liveness prob in Kubernetes (or similar in other platforms), you should notify another side of communication that you're going away (meaning send CloseFrame on terminate event).

If you use Spring Boot hooks for system will be shut down can you close the session there or it is too late.

**Comment by Fetsivalen** [comment]
Hi @violetagg 
I  tried a few combinations and this one as part of example from [this comment](https://github.com/reactor/reactor-netty/issues/1091#issuecomment-631090653)

```
 Mono<Void> serverClosedStatusStream = serverSession.receiveClosedStatus()
 .flatMap(closedStatus -> {
                    if (closedStatus.code() != -1) {
                        return serverSession.close(new CloseStatus(closedStatus.code(), closedStatus.reasonText()));
                    }
                    return serverSession.close(GOING_AWAY);

                });
```
seems to be the closest to the hook.
And with it got next exception in my shutdown test. So Think it's too late
```
Caused by: java.lang.IllegalStateException: executor not accepting a task
	at io.netty.resolver.AddressResolverGroup.getResolver(AddressResolverGroup.java:61) ~[netty-resolver-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.bootstrap.Bootstrap.doResolveAndConnect0(Bootstrap.java:194) [netty-transport-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.bootstrap.Bootstrap.access$000(Bootstrap.java:46) [netty-transport-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:180) [netty-transport-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:166) [netty-transport-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:577) [netty-common-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:551) [netty-common-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:490) [netty-common-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:615) [netty-common-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.util.concurrent.DefaultPromise.setSuccess0(DefaultPromise.java:604) [netty-common-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:104) [netty-common-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:84) [netty-transport-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.safeSetSuccess(AbstractChannel.java:984) [netty-transport-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:504) [netty-transport-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:417) [netty-transport-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:474) [netty-transport-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164) [netty-common-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472) [netty-common-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500) [netty-transport-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-common-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.48.Final.jar:4.1.48.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.48.Final.jar:4.1.48.Final]
	at java.lang.Thread.run(Thread.java:835) [?:?]
```
Maybe it is worth to mention that my project also have "gracefull shutdown"
according to https://github.com/reactor/reactor-netty/issues/1022#issue-5780320620

**Comment by violetagg** [comment]
@Fetsivalen Were you able to resolve this requirement with a shutdown hook?

**Comment by Fetsivalen** [comment]
Hi @violetagg  
For now, in our infrastructure, this requirement was resolved in client implementation (not a good solution actually), and that's why I switched to another stuff.
I'm going to revisit it next week, will let you know once I will dig into it.

**Comment by JamesChenX** [comment]
Hi @violetagg 
I think the status code -1 is error-prone and I encounter a bug related to this ticket.
The code in reactor.netty.http.server.WebsocketServerOperations#sendCloseNow try to create a WebSocketCloseStatus with a status code -1 while Spring WebFlux will just throw an exception "Invalid status code" because it's an invalid status code. As a result, the code in sendCloseNow can never work expectedly if CloseWebSocketFrame is null.

![image](https://user-images.githubusercontent.com/44962329/88729645-f4135b00-d166-11ea-865d-a9362d3e9250.png)
![image](https://user-images.githubusercontent.com/44962329/88729730-24f39000-d167-11ea-86b7-0903fedd700c.png)

**Comment by violetagg** [comment]
@JamesChenX your comment was fixes with #1406

**Comment by violetagg** [comment]
@Fetsivalen Closing this issue. If the suggested solution is not enough we can reopen it.
